asterius/src/Asterius/Internals/Parallel.hs

```diff
@@ -0,0 +1,35 @@
+{-# LANGUAGE LambdaCase #-}
+
+module Asterius.Internals.Parallel
+  ( parallelFor
+  )
+where
+
+import Control.Concurrent
+import Control.Concurrent.MVar
+import Control.Monad
+import Data.IORef
+
+-- | Given the worker thread pool capacity @c@, @parallelFor c xs f@ maps @f@
+-- on @xs@ in parallel on the global thread pool, and concatenates the results.
+-- If @c = 1@ then it is equivalent to @mconcat <$> mapM fn xs@, thus avoiding
+-- threading overhead. NOTE: this function does not only assume associativity
+-- for @<>@ (as promised by the @Monoid@ instance), but also symmetry.
+parallelFor :: Monoid r => Int -> [a] -> (a -> IO r) -> IO r
+parallelFor n xs fn
+  | n >= 2 = do
+    input <- newIORef xs
+    mvars <- replicateM n newEmptyMVar
+    let getNextElem = atomicModifyIORef input $ \case
+          [] -> ([], Nothing)
+          (y:ys) -> (ys, Just y)
+        loop mvar acc = getNextElem >>= \case
```

Comment:
Don't forget to use `!acc`. We want the accumulation to happen in the worker thread, instead of passing a thunk consisting a long chain of `<>` to the parent thread.